sdk/src/main/scala/io/horizen/account/state/ForgerStakeStorageV3.scala

```diff
@@ -0,0 +1,404 @@
+    ForgerMap.getPagedListOfForgers(view, startPos, pageSize)
+
+  def getForger(view: BaseAccountStateView, signKey: PublicKey25519Proposition, vrfKey: VrfPublicKey): Option[ForgerInfoV3] = {
+    val forgerKey = getForgerKey(signKey, vrfKey)
+    ForgerMap.getForger(view, forgerKey)
+  }
+
+  def addStake(view: BaseAccountStateView,
+               signKey: PublicKey25519Proposition,
+               vrfPublicKey: VrfPublicKey,
+               epochNumber: Int,
+               delegatorPublicKey: Address,
+               stakedAmount: BigInteger): Unit = {
+
+    val forgerKey = getForgerKey(signKey, vrfPublicKey)
+    val forgerHistory = StakeHistory(forgerKey)
+    val forgerHistorySize =  forgerHistory.getSize(view)
+    if (forgerHistorySize == 0)
+      throw new ExecutionRevertedException(s"Forger doesn't exist.")
+
+    forgerHistory.updateOrAddCheckpoint(view, forgerHistorySize, epochNumber, latestStake => latestStake.add(stakedAmount))
+
+    val stakeKey = getStakeKey(forgerKey, delegatorPublicKey)
+    val stakeHistory = StakeHistory(stakeKey)
+    val stakeHistorySize = stakeHistory.getSize(view)
+    stakeHistory.updateOrAddCheckpoint(view, stakeHistorySize, epochNumber, latestStake => latestStake.add(stakedAmount))
+    if (stakeHistorySize == 0)
+      addNewDelegator(view, forgerKey, delegatorPublicKey)
+  }
+
+  def removeStake(view: BaseAccountStateView,
+                  blockSignProposition: PublicKey25519Proposition,
+                  vrfPublicKey: VrfPublicKey,
+                  epochNumber: Int,
+                  delegatorPublicKey: Address,
+                  stakedAmount: BigInteger): Unit = {
+
+
+    val forgerKey = getForgerKey(blockSignProposition, vrfPublicKey)
+    val forgerHistory = StakeHistory(forgerKey)
+    val forgerHistorySize = forgerHistory.getSize(view)
+    if (forgerHistorySize == 0)
+      throw new ExecutionRevertedException("Forger doesn't exist.")
+
+    val stakeKey = getStakeKey(forgerKey, delegatorPublicKey)
+    val stakeHistory = StakeHistory(stakeKey)
+    val stakeHistorySize = stakeHistory.getSize(view)
+    if (stakeHistorySize == 0)
+      throw new ExecutionRevertedException(s"Delegator ${BytesUtils.toHexString(delegatorPublicKey.toBytes)} doesn't have stake with the forger.")
+
+
+    def subtractStake(latestStake: BigInteger): BigInteger = {
+      val newAmount = latestStake.subtract(stakedAmount)
+      if (newAmount.signum() == -1)
+        throw new ExecutionRevertedException(s"Not enough stake. Claimed $stakedAmount, available $latestStake")
+      newAmount
+    }
+
+    stakeHistory.updateOrAddCheckpoint(view, stakeHistorySize, epochNumber, subtractStake)
+    forgerHistory.updateOrAddCheckpoint(view, forgerHistorySize, epochNumber, subtractStake)
+  }
+
+  def getAllForgerStakes(view: BaseAccountStateView): Seq[(PublicKey25519Proposition, VrfPublicKey, Address, BigInteger)] = {
```

Comment:
Let's add case class for this response and it can be used in many other places:
`case class ForgerStake(blockSignProposition: PublicKey25519Proposition, vrfPublicKey: VrfPublicKey, rewardAddress: Address, stakedAmount: BigInteger)
`

Reply:
Actually, a class like this exists already (ForgerStakeData). I'll use it, then.